src/main/java/io/gravitee/common/sse/ServerEvent.java

```diff
@@ -0,0 +1,169 @@
+        this(null, data, null, null, null);
+    }
+
+    public boolean isEmpty() {
+        return event == null && data == null && id == null && retry == null;
+    }
+
+    public static ServerEvent parse(@Nullable Buffer buffer) {
+        if (buffer == null) {
+            return new ServerEvent(null, null, null, null);
+        }
+        String event = null;
+        StringBuilder data = new StringBuilder();
+        String id = null;
+        Long retry = null;
+
+        for (Buffer lineBuffer : BufferUtils.split(buffer, new byte[] { '\n' })) {
+            String line = lineBuffer.toString().strip();
+            if (line.startsWith("event:")) {
+                event = line.substring(6).trim();
+            } else if (line.startsWith("data:")) {
+                if (!data.isEmpty()) {
+                    data.append('\n');
+                }
+                data.append(line.substring(5).trim());
+            } else if (line.startsWith("id:")) {
+                id = line.substring(3).trim();
+            } else if (line.startsWith("retry:")) {
+                try {
+                    retry = Long.parseLong(line.substring(6).trim());
+                } catch (NumberFormatException ignored) {
+                    // Per SSE spec, ignore the field if the value is not a valid integer.
+                }
+            }
+        }
+
+        String strData = data.isEmpty() ? null : data.toString();
+
+        return new ServerEvent(event, strData, id, retry, buffer);
+    }
+
+    public Buffer toBuffer() {
+        if (cachedBuffer != null) {
+            return cachedBuffer;
+        }
+
+        StringBuilder buffer = new StringBuilder();
+        if (event != null) {
+            buffer.append("event: ").append(event).append('\n');
+        }
+        if (data != null) {
+            for (String line : data.split("\n")) {
+                buffer.append("data: ").append(line).append('\n');
+            }
+        }
+        if (id != null) {
+            buffer.append("id: ").append(id).append('\n');
+        }
+        if (retry != null) {
+            buffer.append("retry: ").append(retry).append('\n');
+        }
+        return buffer.isEmpty() ? Buffer.buffer() : Buffer.buffer(buffer.append('\n').toString());
+    }
+
+    /**
+     * Create a new ServerEvent with the given data, attempting to preserve the original buffer's structure.
+     * @param data the new data
+     * @return a new ServerEvent instance with the updated data
+     */
+    public ServerEvent withData(String data) {
+        if (Objects.equals(this.data, data)) {
+            return this;
+        }
+
+        if (cachedBuffer == null) {
+            return new ServerEvent(this.event, data, this.id, this.retry, null);
+        }
+
+        List<String> newLines = new ArrayList<>();
+        boolean dataLinesReplaced = false;
+
+        for (Buffer line : BufferUtils.split(this.cachedBuffer, new byte[] { '\n' })) {
```

Comment:
I'm not fully sure to understand the line replacement algorithm. I would have expected a simpler approach to rebuild the event. Can we have a chat?

Reply:
Let's keep it like this 👍